### ISSUE TITLE
Unable to fetch commits (sentry.io + gitlab.com)

### ISSUE THREAD

**Issue by vmk-dev** [getsentry/sentry#34903]
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

1. Log into sentry.io
2. Click on "Organisation settings"
3. Click on "Integrations"
4. Click on "GitLab"
5. Click on "Configurations" tab
6. Click on "Configure" besides your repo
7. You get a 400-Bad Request

### Expected Result

Until some days ago we had no issues but now we get an email from sentry telling us that it could not fetch the commits from GitLab. Used to work perfectly fine and we did not change anything.

### Actual Result

We get a 400 - Bad Request saying:

`detail: "HTTP 400 (invalid_grant): The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."`

I already contacted you in 10/2021 to let you know, that GitLab changed their settings for access tokens with expiration per default. Then I fixed it by ticking this off and you changed your docs regarding this. Since then it worked fine but now we have this problem... Do you know if something changed here again or what could be the reason? It is no fun to have to change the auth token every few months so I hope you guys find a way of consistency with your GitLab integration 🙏

**Comment by getsentry-release** [comment]
Routing to @getsentry/ecosystem for [triage](https://develop.sentry.dev/processing-tickets/#3-triage). ⏲️

**Comment by leeandher** [comment]
Hi @vmk-dev, thanks for your issue report. I've done some investigation into this issue and I believe it was caused by our mishandling of concurrent requests to retrieve a refresh token for a given GitLab integration. We likely, through some error, didn't persist the refresh token we were issued and now need to the user to re-authorize the integration on GitLab's side.

Unfortunately, the authorization flow is heavily coupled with the installation flow, so until we split those up, **we can only recommend that you uninstall and reinstall**. 

In the future, we hope to have a 're-authenticate' workflow, that can preserve the same integration, but issue it new updated credentials if something unexpected like this happens. I've made a ticket to tackle this tech debt in the future, but it'll have to get prioritized before we can provide a timeline.

If other users encounter this problem, please react and provide insights on this thread, as it'll help this refactor get prioritized!

**Comment by MarkusH** [comment]
Hi @leeandher. Thanks for the input. We've been seen the issue as well for an estimated 2 to 3 weeks. At first, I thought it was a temporary error on either Sentry or GitLab.com. But correlating some of these emails with actions taken on GitLab.com, I figured, I go and look if it's a Sentry issue.

**Comment by vmk-dev** [comment]
@leeandher Thank you for investigating. That means for now we have no other choice than re-issuing a new PAT which means the very same problem can reoccur on your site as for now you cannot fix the issue as the logic is coupled too tight?

**Comment by vmk-dev** [comment]
@gee-forr good input I also noticed the new major release v15.x in GitLab. Maybe @leeandher can also have a look on GitLab's changelog as there is a pretty high chance things have changed or been deprecated which you use on your site at sentry.io

**Comment by pierresebastien** [comment]
I'm using Sentry self hosted (version 22.5.0) and GitLab self hosted (version 15.0.0) and it seems indeed related to this [breaking change](https://about.gitlab.com/blog/2022/04/18/gitlab-releases-15-breaking-changes/#oauth-tokens-without-an-expiration) introduced in GitLab 15.
The integrations was broken on my side after upgrading GitLab (when I want to change an existing configuration, I received this error : HTTP 400 (invalid_grant): The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.) so I tried to reconfigure after the upgrade of GitLab and it works for several hours but I have the same issue after some time. According [GitLab documentation](https://docs.gitlab.com/ee/integration/oauth_provider.html#expiring-access-tokens), token expires after 2hours now.

**Comment by marvinatgithub** [comment]
> If other users encounter this problem, please react and provide insights on this thread, as it'll help this refactor get prioritized!

I received _Unable to Fetch Commits_ emails since 2022-05-17. In Sentry's GitLab integration config I also received 
```
{"detail":"HTTP 400 (invalid_grant): The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."}
```
from `https://sentry.io/api/0/organizations/.../integrations/.../repos/`. We do not self host neither GitLab nor Sentry.

Yesterday I [reinstalled](https://sentry.io/organizations/aym/integrations/gitlab/setup/?) the GitLab integration (checking only 'Confidential' in 4.) as recommended by @leeandher what fixed it since then.

**Comment by gee-forr** [comment]
@marvinatgithub - can you confirm that your reinstalled gitlab integration is still working 24 hours after setting it up? We've all reinstalled the integration, and it works, but only for 2 hours before expiring again.

**Comment by vmk-dev** [comment]
@gee-forr Yes, I can confirm that. We also re-issued a new token and setup the gitlab integration in sentry.io from 0. It works since then (like a week now) and hopefully will continue to work.

**Comment by usrme** [comment]
Am I right in understanding that when a configuration gets uninstalled then absolutely all the data (discussions, events, projects) gets removed or is there some other way to reinstall the configuration? Using self-hosted GitLab 15.0.0 and managed Sentry.

**Comment by vmk-dev** [comment]
@usrme no this was not the case. We use gitlab.io and sentry.io and in our case nothing was lost. Your projects and issues are still existing in sentry and even the interlinking between sentry issues and gitlab issues was correctly there. So in fact we did not lose any data. But we use GitLab for discussions, labels, assignments, etc. so that kind of data may be lost in sentry, but I do not know as we do not use that.

**Comment by arusahni** [comment]
As someone who was recently burned by this, would it be possible for Sentry to do any of the following?

- Add a GitLab section to this help page (that's linked to from the unable to fetch commits error email): https://help.sentry.io/product-features/integrations/why-am-i-receiving-the-email-unable-to-fetch-commits/
- Proactively message all sentry.io orgs that have GitLab integrations enabled?

**Comment by Vonter** [comment]
I had the same issue on a self-hosted Sentry and self-hosted GitLab 15.0 setup. Reconfiguring/Reinstalling the GitLab integration would fix the issue but it would break again in 2 hours. Eventually I [found this fix](https://github.com/dexidp/dex/issues/2316#issuecomment-1143663212) that worked:
* Edit `opt/gitlab/embedded/lib/ruby/gems/2.7.0/gems/doorkeeper-5.5.0.rc2/lib/doorkeeper/config.rb` and set the value of `access_token_expires_in` to a long enough time period (or use `nil` for a token that never expires).
* Restart Gitlab with `gitlab-ctl restart` to apply the access token expiry configuration change

Since I applied this fix, the integration hasn't broken again

**Comment by bkcsfi** [comment]
@leeandher  

> We are looking to address this soon with a re-authorize flow, to preserve your integrations, but can’t provide a timeline just yet, though it is a priority for https://github.com/orgs/getsentry/teams/ecosystem.

Is there an existing issue/fix for avoiding the `access_token_expires_in` workaround suggested by @Vonter ?

I really don't want to have to modify/patch the gitlab code every time I pull a deploy a gitlab release on-prem. 

I'm not worried about preserving past integrations, but I do want the integrations to continue to work for more than 2 hours after I re-configure them them all.

Thanks

**Comment by haiddablitz** [comment]
`access_token_expires_in` workaround suggested by @Vonter does not work for me. Was able to reconnect Sentry but the next day the connection was broken again.

**Comment by leeandher** [comment]
The current version of Sentry does have functionality for handling refresh tokens, so the integration should no longer be failing after two hours. If that is occurring, you may want to create a separate issue for that to get it prioritized or look into your GitLab configuration to make sure it matches our docs/guides.

Unfortunately the refactor I'd mentioned previously hasn't taken place yet, so for now we can still only suggest uninstalling and reinstalling to fix your integrations if they've been expired when migrated to GitLab 15.0. We also haven't taken a look at implementing a patch similar to what @Vonter suggested yet, but please feel free to make a PR if you'd like to see it added to the core Sentry instead of having to manually patch your Self-Hosted instances each time you update!

**Comment by arathunku** [comment]
I come here to confirm @haiddablitz's comment. We've delayed GitLab v15 update for a bit, but had to finally update this week. I was expecting problems with the integration and set up it from scratch, following the docs. I've even created a new app in GitLab, authorized successfully and everything worked yesterday. I've restored all repos. Today, there's exactly **this** issue. It's unable to fetch commits and when I try to access "Repositories" page, there's an error. We use sentry.io, so it looks like there's some kind of bug in the current version of Sentry.

org name: `gat`

![image](https://user-images.githubusercontent.com/749393/183892240-2a522a7f-c616-4416-9954-3ece5c9d75de.png)

**Comment by e-jjj** [comment]
> Thanks for directing us to the GitLab 15.0 changes, that is likely the culprit for the recent surge in this issue. We’re still trying to understand if there was anything Sentry could have done to migrate the ‘no-expiration’ GitLab apps from pre-14.3 to being refreshed after this GitLab 15.0 update, but it appears those those applications are the ones seeing this token refresh failure error.
> 
> That said, my [previous comment](https://github.com/getsentry/sentry/issues/34903#issuecomment-1136533582) about the installation/authentication issue still stands. Unfortunately at the moment we can only recommend you uninstall and reinstall your GitLab configuration, but doing so will remove any linked issues, repositories, codeowners files, code mappings and user/team mappings. Your projects, issues and other Sentry data will remain, but anything related to GitLab will be removed.
> 
> We are looking to address this soon with a re-authorize flow, to preserve your integrations, but can’t provide a timeline just yet, though it is a priority for @getsentry/ecosystem.

@leeandher  Has there been any progress with a solution on the integration with gilab?  I'm running sentry 21.8 (self hosted)  and gitlab 15 (gitlab.com) and facing the oauth token expiration every 2 hours even after re-creating the integrations. Any ideas would be appreciated. Thanks!

**Comment by bkcsfi** [comment]
@leeandher I have the same concerns as @e-jjj  except I am a paying sentry.io customer

I don't mind re-doing the setup, but I only want to re-do the setup only once.

The value of my sentry.io subscription has been diminished due to this unsolved issue that's been open since May.

Very frustrating.

**Comment by haiddablitz** [comment]
Any updates on this issue? We use self hosted Gitlab (version 15.2.4) and self hosted Sentry (version 22.8.0) and still receive "unable to fetch commits" errors.

**Comment by yaboeing** [comment]
Hi folks!
Sentry 21.12.0 14f603c and Gitlab 15.2.2-ee. Integration works as expected even after the token expiration period.
Seems the fix was [here](https://github.com/getsentry/sentry/pull/29747/files#diff-2d720bd6fa142076c64674597f44a95a5edef934e20544c8332346b4912f3b79R336)

**Comment by flyinbutrs** [comment]
Has anyone figured out if this is live in sentry.io? I've been holding off on upgrading our self hosted gitlab to 15 because I'd read about this issue.

**Comment by victorelec14** [comment]
> Any updates on this issue? We use self hosted Gitlab (version 15.2.4) with self hosted Sentry (version 22.8.0) and still receive "unable to fetch commits" errors.

related with getsentry/self-hosted#1672 ??

**Comment by armenzg** [comment]
Hello,
I have been inquiring about this and what the current state is.

If you take these steps you will get into a working state:
* Uninstall the Gitlab integration from Sentry
* Delete the Sentry Application you created in Gitlab
* Follow the Gitlab integration steps to install it

There's a caveat, if you have Gitlab issues linked to Sentry you will loose that data. This is unfortunate but we have not been able to prioritize the work to re-authenticate and we currently don't have the bandwidth to investigate.

Our apologies for the delay and that the upgrade path looses you some data. We wish we have had the bandwidth to have a better story here and sadly we won't have it for the foreseeable future.

As a recap:
- Going from GL 14 to 15 added a breakage where tokens did not have an [expiration data](https://about.gitlab.com/blog/2022/04/18/gitlab-releases-15-breaking-changes/#oauth-tokens-without-an-expiration)
- Sentry 21.12.0 added support for refreshing tokens ([fix](https://github.com/getsentry/sentry/pull/29747/files#diff-2d720bd6fa142076c64674597f44a95a5edef934e20544c8332346b4912f3b79R336))

**Comment by armenzg** [comment]
Something came to my mind that it may work.

For those with self-hosting, you can search for your Gitlab integration in Sentry (21.12.x and newer) and modify the `metadata` JSON object. Put in `webhook_secret` the token from Gitlab. This value comes from Gitlab as the HTTP header `X-Gitlab-Token`. Once you update secret in Sentry it should work.

For those in SaaS, if someone feels like coding, there's a chance to change the Gitlab integration page to allow you to set a new token for the integration. We may be able to put this behind an Early Adopters flag. Probably somewhere around here:
<img width="672" alt="image" src="https://user-images.githubusercontent.com/44410/193136012-1be9e9a2-fff2-4637-a499-ca0aaef6325e.png">

If someone is feeling adventurous and would like to give this a try it would be immensely helpful. If this works, it may be possible to put a bit of time before end of year to have at least this minimal fix.

**Comment by Christophe31** [comment]
Where should we find this `metadata` JSON object?

**Comment by armenzg** [comment]
> Where should we find this `metadata` JSON object?

Hi @Christophe31 you can use `sentry shell` to import the model and the query the DB:

`sentry shell -c "from sentry.models.integrations import Integration; print(Integration.objects.get(provider='gitlab').metadata['access_token'])"`

**Comment by jerbob92** [comment]
I'm running GitLab 15.7.0 and Sentry 21.12.0 and this issue still seems to happen for me after reinstalling the GitLab integration (it did work for a short while). It results in a 400 error on the GitLab side and Sentry generates the error `The webhook secrets do not match.`.

**Comment by ctgdevops** [comment]
we are running GitLab 15.4.0 and Sentry 22.8.0 and we are still receiving _Unable to Fetch Commits_ errors in email from our Sentry instance.

**Comment by donaldpipowitch** [comment]
We also see this problem with gitlab.io. We occasionally see _"Unauthorized: either your access token was invalid or you do not have access"_, but we setup the whole integration again as described in the docs. We asked the support chat and they said we should raise our problem here.

**Comment by spawnia** [comment]
Seeing that in principle the refresh token flow is supported now, I believe that the remaining issue might just be with self-hosted instances that integrate with GitLab instances using self-signed certificates. The refresh token workflow does not respect the setting `Verify SSL` made during setup of the integration, so it fails to refresh expired tokens. I reported this issue in https://github.com/getsentry/sentry/issues/46491 and attempted a fix in https://github.com/getsentry/sentry/pull/46492.

**Comment by isra17** [comment]
I also had this error in sentry with and cloud gitlab.com instance. I did reset our integration in Sentry and the error is coming back every once in a while. It seems to be intermittent, once every few hundreds commit I would say?

**Comment by Codelica** [comment]
Could someone confirm this issue (refreshing the Gitlab token properly) is fixed in Sentry 21.12.0+ ?

I'm not concerned about the potential self-signed cert issue mentioned, but it's hard to tell if this issue was resolved  considering it's still open.  

Thanks!

**Comment by haiddablitz** [comment]
Can confirm the Gitlab token is properly refreshed in self hosted Sentry with version 23.8.0

**Comment by hubertdeng123** [comment]
Thank for confirming, I’m going to go ahead and close this then.
